src/Glpi/Asset/AssetDefinition.php

```diff
@@ -346,41 +347,24 @@ public function post_updateItem($history = true)
         parent::post_updateItem();
 
         if (in_array('capacities', $this->updates)) {
-            $new_capacities = @json_decode($this->fields['capacities']);
-            $old_capacities = @json_decode($this->oldvalues['capacities']);
-
-            if (!is_array($new_capacities)) {
-                // should not happen, do not trigger cleaning to prevent unexpected mass deletion of data
-                trigger_error(
-                    sprintf('Invalid `capacities` value `%s`.', $this->fields['capacities']),
-                    E_USER_WARNING
-                );
-                return;
-            }
-            if (!is_array($old_capacities)) {
-                // should not happen, do not trigger cleaning to prevent unexpected mass deletion of data
-                trigger_error(
-                    sprintf('Invalid `capacities` value `%s`.', $this->oldvalues['capacities']),
-                    E_USER_WARNING
-                );
-                return;
-            }
+            $new_capacities = $this->getDecodedCapacities();
+            $old_capacities = $this->getDecodedCapacities($this->oldvalues['capacities']);
 
-            $added_capacities = array_diff($new_capacities, $old_capacities);
-            foreach ($added_capacities as $capacity_classname) {
-                $this->onCapacityEnabled($capacity_classname);
+            $added_capacities = array_diff_key($new_capacities, $old_capacities);
+            foreach ($added_capacities as $capacity) {
+                $this->onCapacityEnabled($capacity->getName(), $capacity->getConfig());
             }
 
-            $removed_capacities = array_diff($old_capacities, $new_capacities);
-            foreach ($removed_capacities as $capacity_classname) {
-                $this->onCapacityDisabled($capacity_classname);
+            $removed_capacities = array_diff_key($old_capacities, $new_capacities);
+            foreach ($removed_capacities as $capacity) {
+                $this->onCapacityDisabled($capacity->getName());
```

Comment:
You should pass the capacity instance itself, we may need to adapt the behaviour depending on the configuration.